### PR TITLE
feat: forget password feature

### PR DESCRIPTION
**Pull request by Harish-osmosys**
## Email SS:

image

## Summary by CodeRabbit

- **New Features**
    - Introduced email notification functionality, enabling the sending of password reset emails.
    - Added support for custom email templates for specific events like "ForgetPassword".
- **Enhancements**
    - Improved token management for email actions with specified token lifespan settings.
    - Enhanced email sending capabilities with HTTP client configuration.
- **Refactor**
    - Updated user management to integrate email notifications for password reset processes.
- **Chores**
    - Updated project dependencies, including adding `Newtonsoft.Json` version `13.0.3`.
    - Added `.vscode` folder to `.gitignore` to improve developer experience.

### CODE REVIEW

File: DotnetFoundation/DotnetFoundation.Application/Interfaces/Persistence/IEmailRepository.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace DotnetFoundation.Application.Interfaces.Persistence;
+
+public interface IEmailRepository
+{
+    public Task<string> SendForgetPasswordEmailAsync(string email, string body);
```
[the remark this answers is not in the document]
Comment (sameer-s-b) [reply DotnetFoundation/DotnetFoundation.Application/Interfaces/Persistence/IEmailRepository.cs]: This needs to be resolved.

File: DotnetFoundation/DotnetFoundation.Application/Interfaces/Services/IEmailService.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace DotnetFoundation.Application.Interfaces.Services;
+
+public interface IEmailService
+{
+    public Task<string> SendEmail(string emailId, string subject, string body);
```
[the remark this answers is not in the document]
Comment (sameer-s-b) [reply DotnetFoundation/DotnetFoundation.Application/Interfaces/Services/IEmailService.cs]: Should be resolved

File: DotnetFoundation/DotnetFoundation.Application/Services/EmailService/EmailService.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    private readonly IEmailRepository _emailRepository;
+
+    public EmailService(IEmailRepository emailRepository)
+    {
+        _emailRepository = emailRepository;
+    }
+    public async Task<string> SendEmail(string emailId, string subject, string body)
+    {
+        string res = await _emailRepository.SendForgetPasswordEmailAsync(emailId, body).ConfigureAwait(false);
+        return res;
```
[the remark this answers is not in the document]
Comment (sameer-s-b) [reply DotnetFoundation/DotnetFoundation.Application/Services/EmailService/EmailService.cs]: This needs to be fixed.

File: DotnetFoundation/DotnetFoundation.Domain/Entities/Notification.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+{
+    SMTP = 1,
+    MailGun = 2,
+    WhatsApp = 3
+}
+public enum EmailEvents
+{
+    ForgetPassword
+}
+public class NotificationData
```
Comment (sameer-s-b) [inline DotnetFoundation/DotnetFoundation.Domain/Entities/Notification.cs:15]: Have spaces in between class blocks.

File: DotnetFoundation/DotnetFoundation.Domain/Entities/Notification.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+public static class EmailConfig
+{
+    public static Dictionary<EmailEvents, EmailTemplate> EmailTemplatesDictionary = new()
+    {
+        {
+            EmailEvents.ForgetPassword,
+            new EmailTemplate
+            {
+                TemplatePath = "../DotnetFoundation.Domain/Templates/Emails/ForgetPasswordTemplate.html",
```
Comment (sameer-s-b) [inline DotnetFoundation/DotnetFoundation.Domain/Entities/Notification.cs]: Email templates can come from appSettings.json file.

```
Emails: {
	"forgetPassword": {
		"path": "",
		"subject": ""
	},
	
}
```

File: DotnetFoundation/DotnetFoundation.Api/appsettings.json [hunk 6]
```diff
@@ -9,9 +9,14 @@
     "DBConnection": "server=foundation-mariadb;Port=3306;database=DotnetFoundation;Uid=root;pwd=Change123"
   },
   "Jwt": {
-    "Key": "My-Super-Secret-Key!@#$%",
+    "Key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQC2CoxBJo5TR2O2agbJ3rvoQAC1CGNOfcGWlyxzUcIfoXBLoDjimzfb/nIfvku3m4ZTzsx88+mJw4gtNo3xx3T4nVszGTOEeTYZod9WlekilKtWMCCYmUrPbdCOVsIgiD4om3yDvoIarVd28PymS7Ub81rYhMFsm+Pmq3Ms5l/VXkSdyh8D60Z6KjDQh39EVs6+qzY3PJHNnZgLx9ivz40mCmOkYd2e58U/z0kX20APr/+zJZtZJm15uFCm4C3aRzQ78lQ7cZtt3pE3Bsth1wS4/utwYWm930YCNMqZeYBNn0n6qfTOkS/wDFJpKNgC3+7v4THoqMuzAO/w7hKDsjSP4vJ2bm1tnoB0yEjkK2SMT85wNxWl7JkIRWW1FqxdWMQD5c1r3g5TvAzfPdJ/rw5zZNIUyq2r/7lHJrO1UGEsmSddOvBEk4uQ4J905xPgxT8w3Grbz0QjHvZh+AoYRu988b5Nc/3NOVf8EWqnOf55vRW9lxU51r40dpcQPkBik3Q0F1r1VSpPHvr0H0S+aSa+dgim4fnzA/mEcD/x2uUMQP0RbEHnsU/dR7fMtSqUVEDKD9nlvg/TW2lWme/sW2L/Dg1//bUODrdgtimEjBbWB+Ce9HG/+Os5QaFqp9Fgr9ADGlMOFlLCja8CkR+e0DEcJ+9kf7pl0dsy2XzfXUUr6Q== work@Windows",
```
Comment (sameer-s-b) [inline DotnetFoundation/DotnetFoundation.Api/appsettings.json]: Create an .env file and use it to replace the text in the appSettings.json file in the Dockerfile.

File: DotnetFoundation/DotnetFoundation.Domain/Entities/Notification.cs (hunk 4 above)
Comment (Harish-osmosys) [reply DotnetFoundation/DotnetFoundation.Domain/Entities/Notification.cs:15]: fixed

File: DotnetFoundation/DotnetFoundation.Application/Interfaces/Services/IEmailService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply DotnetFoundation/DotnetFoundation.Application/Interfaces/Services/IEmailService.cs]: fixed

File: DotnetFoundation/DotnetFoundation.Application/Interfaces/Persistence/IEmailRepository.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply DotnetFoundation/DotnetFoundation.Application/Interfaces/Persistence/IEmailRepository.cs]: fixed

File: DotnetFoundation/DotnetFoundation.Application/Services/EmailService/EmailService.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply DotnetFoundation/DotnetFoundation.Application/Services/EmailService/EmailService.cs]: fixed

File: DotnetFoundation/DotnetFoundation.Domain/Entities/Notification.cs (hunk 5 above)
Comment (Harish-osmosys) [reply DotnetFoundation/DotnetFoundation.Domain/Entities/Notification.cs]: updated

File: DotnetFoundation/DotnetFoundation.Infrastructure/Persistence/UserRepository.cs [hunk 7]
```diff
@@ -129,14 +132,20 @@ public async Task<string> ForgotPasswordAsync(string email)
     {
         IdentityApplicationUser user = await _userManager.FindByEmailAsync(email).ConfigureAwait(false) ?? throw new Exception("Invalid Email");
         string token = await _userManager.GeneratePasswordResetTokenAsync(user).ConfigureAwait(false);
-        return token;
+        await _emailRepo.SendForgetPasswordEmailAsync(email, token).ConfigureAwait(false);
+
+        return "Success";
```
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply DotnetFoundation/DotnetFoundation.Infrastructure/Persistence/UserRepository.cs]: updated

File: DotnetFoundation/DotnetFoundation.Api/appsettings.json (hunk 6 above)
Comment (Harish-osmosys) [reply DotnetFoundation/DotnetFoundation.Api/appsettings.json]: resolved

**Review by sameer-s-b — APPROVED** [review]

**Pull request merged**